### PR TITLE
Another take at enabling Pallas GPU tests on x64

### LINKED ISSUE #18847: [Pallas Triton] Reduce Arg OP - Not working in X64 mode
### Description

PALLAS/Triton Lowering rules do not work with X64 mode `jax.config.update("jax_enable_x64", True)` activated:

### What jax/jaxlib version are you using?

Jax Nightly

### Which accelerator(s) are you using?

GPU

### Additional system info?

_No response_

### NVIDIA GPU info

```python
import jax
from jax.experimental import pallas as pl
import jax.numpy as jnp

jax.config.update("jax_enable_x64", True)  # Enable INT64  - If you comment - Everything works

# ================ NORMAL JAX ================ #
x = jnp.arange(8).astype(jnp.float32)
print(f"{jnp.argmax(x).dtype=}")   # jnp.argmax(x).dtype=dtype('int64')

# ================ PALLAS CALL ================ #
def argmax_kernel(x_ref, o_ref):
    x = pl.load(x_ref, ())
    out = jnp.argmax(x)
    # print(out.dtype)  # Shows INT64 if uncommented
    pl.store(o_ref, (), out)

@jax.jit
def exec_fn(_x):
    return pl.pallas_call(
        argmax_kernel, 
        out_shape=jax.ShapeDtypeStruct(
            (), 
            dtype=(
                jnp.int64 
                if jax.config.values["jax_enable_x64"] else 
                jnp.int32
            )
        ),  # One single "index_dim" in INT64
        grid=1
    )(_x)

x = jnp.arange(8).astype(jnp.float32)

print(f"{exec_fn(x).dtype=}")   # FAIL
```

**Error:**
```python
File /usr/local/lib/python3.10/dist-packages/jax/_src/pallas/triton/lowering.py:1009, in _argreduce_lowering(body, ctx, a, axes, index_dtype)
   1008 if index_dtype != jnp.int32:
-> 1009   raise ValueError("`index_type` must be i32.")

ValueError: `index_type` must be i32.
```

It produces the following `jaxpr`:

```
{ lambda ; a:Ref{float32[8]} b:Ref{int64[]}. let
    c:f32[8] <- a[:]
    d:i64[] = argmax[axes=(0,) index_dtype=int64] c
    b[] <- d
  in () }
{ lambda ; a:f32[] b:i64[] c:f32[] d:i64[]. let
    e:bool[] = gt a c
    f:bool[] = lt a c
    g:i64[] = min b d
    h:i64[] = pjit[
      jaxpr={ lambda ; i:bool[] j:i64[] k:i64[]. let
          l:i64[] = select_n i k j
        in (l,) }
      name=_where
    ] f d g
    m:i64[] = pjit[
      jaxpr={ lambda ; i:bool[] j:i64[] k:i64[]. let
          l:i64[] = select_n i k j
        in (l,) }
      name=_where
    ] e b h
    n:f32[] = max a c
  in (n, m) }
{ lambda ; a:bool[] b:i64[] c:i64[]. let d:i64[] = select_n a c b in (d,) }
{ lambda ; a:bool[] b:i64[] c:i64[]. let d:i64[] = select_n a c b in (d,) }
```

---------------

Quite obviously if you remove `jax.config.update("jax_enable_x64", True)` then everything falls back to INT32 and consequently works

### PR DESCRIPTION
**Pull request by superbobry**
Closes #18847.

### CODE REVIEW

File: jax/_src/lax/control_flow/for_loop.py [hunk 1]
```diff
@@ -132,7 +133,9 @@ def wrapped_body(i, refs):
   nsteps, = nsteps
   flat_state, state_tree = tree_flatten(init_state)
   state_avals = map(state_utils.val_to_ref_aval, flat_state)
-  idx_aval = core.ShapedArray((), jnp.dtype("int32"))
+  idx_aval = core.ShapedArray(
+      (), jnp.int64 if config.enable_x64.value else jnp.int32
```
Comment (gnecula) [inline jax/_src/lax/control_flow/for_loop.py:137]: You could use `dtypes.canonicalize_dtype` which contains that conditional (and is actually more likely to work when we use the x64 context manager)

File: jax/_src/pallas/triton/lowering.py [hunk 2]
```diff
@@ -2606,6 +2607,10 @@ def _ir_constant(v: object, t: ir.Type) -> ir.Value:
   raise NotImplementedError
 
 
+def _i_constant(v: int) -> ir.Value:
```
Comment (gnecula) [inline jax/_src/pallas/triton/lowering.py:2610]: The name seems overly general. iiuc, this is only for loop indices?

File: tests/pallas/pallas_test.py [hunk 3]
```diff
@@ -1730,16 +1707,9 @@ def softmax_kernel(x_ref, y_ref):
 class PallasCallAutodifferentiationInterpreterTest(PallasCallAutodifferentiationTest):
   INTERPRET = True
 
-  def setUp(self):
```
Comment (gnecula) [inline tests/pallas/pallas_test.py:1733]: Thanks!

**Review by gnecula — APPROVED** [review]

File: jax/_src/lax/control_flow/for_loop.py (hunk 1 above)
Comment (superbobry) [reply jax/_src/lax/control_flow/for_loop.py:137]: Done.

File: jax/_src/pallas/triton/lowering.py (hunk 2 above)
Comment (superbobry) [reply jax/_src/pallas/triton/lowering.py]: Inlined into the call site.

**Pull request merged**